ambry-router/src/main/java/com.github.ambry.router/GetBlobInfoOperation.java

```diff
@@ -0,0 +1,243 @@
+  protected final Map<Integer, GetRequestInfo> correlationIdToGetRequestInfo = new TreeMap<Integer, GetRequestInfo>();
+
+  /**
+   * Construct a GetBlobInfoOperation
+   * @param routerConfig the {@link RouterConfig} containing the configs for get operations.
+   * @param clusterMap the {@link ClusterMap} of the cluster
+   * @param responseHandler the {@link ResponseHandler} responsible for failure detection.
+   * @param blobIdStr the blob id associated with the operation in string form.
+   * @param futureResult the future that will contain the result of the operation.
+   * @param callback the callback that is to be called when the operation completes.
+   * @param time the Time instance to use.
+   * @throws RouterException if there is an error with any of the parameters, such as an invalid blob id.
+   */
+  GetBlobInfoOperation(RouterConfig routerConfig, ClusterMap clusterMap, ResponseHandler responseHandler,
+      String blobIdStr, FutureResult<BlobInfo> futureResult, Callback<BlobInfo> callback, Time time)
+      throws RouterException {
+    super(routerConfig, clusterMap, responseHandler, blobIdStr, futureResult, callback, time);
+    operationTracker = new SimpleOperationTracker(routerConfig.routerDatacenterName, blobId.getPartition(), true,
+        routerConfig.routerGetSuccessTarget, routerConfig.routerGetRequestParallelism);
+  }
+
+  /**
+   * Return the {@link MessageFormatFlags} to associate with a getBlobInfo operation.
+   * @return {@link MessageFormatFlags#BlobInfo}
+   */
+  @Override
+  MessageFormatFlags getOperationFlag() {
+    return MessageFormatFlags.BlobInfo;
+  }
+
+  /**
+   * For this operation, create and populate get requests (in the form of {@link RequestInfo}) to send out.
+   * @param requestRegistrationCallback the {@link RequestRegistrationCallback} to call for every request that gets
+   *                                    created as part of this poll operation.
+   */
+  @Override
+  void poll(RequestRegistrationCallback<GetOperation> requestRegistrationCallback) {
+    //First, check if any of the existing requests have timed out.
+    Iterator<Map.Entry<Integer, GetRequestInfo>> inFlightRequestsIterator =
+        correlationIdToGetRequestInfo.entrySet().iterator();
+    while (inFlightRequestsIterator.hasNext()) {
+      Map.Entry<Integer, GetRequestInfo> entry = inFlightRequestsIterator.next();
+      if (time.milliseconds() - entry.getValue().startTimeMs > routerConfig.routerRequestTimeoutMs) {
+        operationTracker.onResponse(entry.getValue().replicaId, false);
+        setOperationException(
+            new RouterException("Timed out waiting for responses", RouterErrorCode.OperationTimedOut));
+        inFlightRequestsIterator.remove();
+      } else {
+        // the entries are ordered by correlation id and time. Break on the first request that has not timed out.
+        break;
+      }
+    }
+
+    checkAndMaybeComplete();
+    if (isOperationComplete()) {
+      return;
```

Comment:
Personal preference indeed, but since you have brought this up earlier as well and since we are not just talking about this case, let me see if I can convince you against rigidly following the practice of returning at the end :)

First of all, in a language like java, methods exit from the middle all the time because of exceptions. The only way to guarantee exit points is by wrapping the whole thing in a `try-finally` block. I do agree that it is not desirable to have methods return from here and there in general, but sometimes it makes sense to return immediately when done. Examples:

- When conditions are met within nested loops:
```
void func() {
...
for (...) {
   for (...) {
      if (condition_is_met) {
        return;
      }
   }
}
...
}

```
Avoiding the return in such a method would require making the code more complicated than it has to be. You would need to set some flag, then break, then have the outer loop check and then break, possibly have the check again after the loop, etc.

- Another example is when there are certain cases that can be disposed off immediately:

```
void func() {
   if (!isOpen) {
      return;
   }
   ...
}
```
Doing it this way lets you keep the core functionality outside of an indentation which makes it more readable in my opinion.

Once again, I am not recommending that we rigidly follow any practice. I think that this should be left to the judgement of the programmer, and that such judgements should be reviewed independently.

As for the case in question, I have moved the first few statements into a method as Siva suggested, and I think the return should be more readable now. If not, let me know.